messaging/activity/FirebaseMessagingActivityGenerator.cs

```diff
@@ -0,0 +1,172 @@
+"    // If we do not have a 'from' field this intent was not a message and should not be handled. It",
+"    // probably means this intent was fired by tapping on the app icon.",
+"    Bundle extras = intent.getExtras();",
+"    if (extras == null) {{",
+"      return;",
+"    }}",
+"    String from = extras.getString(EXTRA_FROM);",
+"    String messageId = extras.getString(EXTRA_MESSAGE_ID_KEY);",
+"    if (messageId == null) {{",
+"      messageId = extras.getString(EXTRA_MESSAGE_ID_KEY_SERVER);",
+"    }}",
+"    if (from != null && messageId != null) {{",
+"      Intent message = new Intent(this, MessageForwardingService.class);",
+"      message.setAction(MessageForwardingService.ACTION_REMOTE_INTENT);",
+"      message.putExtras(intent);",
+"      message.setData(intent.getData());",
+"      MessageForwardingService.enqueueWork(this, message);",
+"    }}",
+"    setIntent(intent);",
+"  }}",
+"",
+"  /**",
+"   * Dispose of the mUnityPlayer when restarting the app.",
+"   *",
+"   * <p>This ensures that when the app starts up again it does not start with stale data.",
+"   */",
+"  @Override",
+"  protected void onCreate(Bundle savedInstanceState) {{",
+"    if (mUnityPlayer != null) {{",
+"      mUnityPlayer.quit();",
+"      mUnityPlayer = null;",
+"    }}",
+"    super.onCreate(savedInstanceState);",
+"  }}",
+"}}"
+  };
+  private readonly string BaseActivityClass = "UnityPlayerActivity";
+  private readonly string BaseGameActivityClass = "UnityPlayerGameActivity";
+
+  private readonly string GeneratedFileTag = "FirebaseMessagingActivityGenerated";
+  // If this tag is present on the generated file, it will not be replaced.
+  private readonly string PreserveTag = "FirebasePreserve";
+
+  private readonly string OutputPath = Path.Combine("Plugins", "Android");
+  private readonly string OutputFilename = "MessagingUnityPlayerActivity.java";
+
+  public int callbackOrder { get { return 0; } }
+  public void OnPreprocessBuild(BuildReport report) {
+    // Only run this logic when building for Android.
+    if (EditorUserBuildSettings.activeBuildTarget != BuildTarget.Android) {
+      return;
+    }
+
+    // Check if the file has already been generated.
+    string[] oldAssetGuids = AssetDatabase.FindAssets("l:" + GeneratedFileTag);
+    if (oldAssetGuids != null && oldAssetGuids.Length > 0) {
+      if (oldAssetGuids.Length != 1) {
+        Debug.LogWarning("FirebaseMessagingActivityEditor found multiple generated files with the label: " +
+                         GeneratedFileTag + " \nNo changes will be made.");
+        return;
+      }
+      string oldAssetPath = AssetDatabase.GUIDToAssetPath(oldAssetGuids[0]);
+      Object oldAsset = AssetDatabase.LoadMainAssetAtPath(oldAssetPath);
+      // If the generated file has been tagged to be preserved, don't change it.
+      string[] labelList = AssetDatabase.GetLabels(oldAsset);
+      if (labelList.Contains(PreserveTag)) {
+        return;
+      }
+      // Delete the old asset.
```

Comment:
I think we should also check if the content of the file is the same to what the code will inject.

- We should reduce the frequency to trigger asset database change as much as possible since that can spin up a whole lot of other process reacting to asset database changes.  
- If the new file is using a different GUID, this is also pretty annoying for version controlling.
- This is also a good chance to check if the user changed the `GameActivity` flag.

Reply:
That's a good point.  Done.